README.rst
flooding
==========================================

Introduction

Usage, etc.

More details in src/flooding/USAGE.txt .


Works with RabbitMQ version 2.8.7.


Install Packages Ubuntu 14.04
-----------------------------
    $ sudo apt-get install binutils libproj-dev python-gdal gdal-bin python-scipy python-numpy


Install production / staging server
-----------------------------------

Linux task machine (i.e. task 200). The task server checks out the
master trunk of flooding.

Init
    $ bin/fab staging_taskserver init
Update
    $ bin/fab staging_taskserver update_task

Init
    $ bin/fab production_taskserver init
Update
    $ bin/fab production_taskserver update_task

See staging-task-200.cfg as an example, it actually serves tasks 200,
210 and 220.

Problems can arise when installing netcdf4. Try:

    $ sudo apt-get install libhdf5-serial-dev libnetcdf-dev

Set /tmp directory on separate divice and check it periodically


Deploying production webserver
--------------------------------

Deploying is done with ansible::

   $ bin/ansible-playbook ansible/deploy.yml --limit production_web


In case you need to release manually: ssh to the relevant server (look in
``ansible/inventory`` for the hostname) and **sudo to buildout**::

    $ cd /srv/HOSTNAME  # different for production/staging
    $ git pull
    $ bin/develop up
    $ bin/buildout
    $ bin/django migrate
    $ bin/supervisorctl reload


WARNING: buildout run on jupiter
--------------------------------

Buildout chokes if you run it with the smb share still mounted.  So,
as root, first unmount ('umount') the share::

  #> umount /srv/flooding.lizardsystem.nl/var/external_data

After buildout finished correctly, mount it again as root::

  #> mount /srv/flooding.lizardsystem.nl/var/external_data


Development installation
------------------------

The first time, you'll have to run the "bootstrap" script to set up setuptools
and buildout::

    $> python bootstrap.py

And then run buildout to set everything up::

    $> bin/buildout

(On windows it is called ``bin\buildout.exe``).

You'll have to re-run buildout when you or someone else made a change in
``setup.py`` or ``buildout.cfg``.

The current package is installed as a "development package", so
changes in .py files are automatically available (just like with ``python
setup.py develop``).

If you want to use trunk checkouts of other packages (instead of released
versions), add them as an "svn external" in the ``local_checkouts/`` directory
and add them to the ``develop =`` list in buildout.cfg.

Tests can always be run with ``bin/test`` or ``bin\test.exe``.



Workflows
------------------------
The next workflow_templates are created on migration:

DEFAULT_TEMPLATE_CODE = 1 (workflow for a scenario with sobek model)
IMPORTED_TEMPLATE_CODE = 2 (workflow for a scenario with onknown model via import)
THREEDI_TEMPLATE_CODE = 3 (workflow for scenario with 3di model)
MAP_EXPORT_TEMPLATE_CODE = 4 (workflow for map's export)

The range of template's code 0 - 50 area reserved for auto workflows.


Upload/download wateren- and kerigne-sahpes
-------------------------------------------
Create a symbolic link ``BUILDOUT_DIR/var/ror_export`` to the mounted directory
(see ``ROR_KERINGEN_PATH`` in ``settings.py``)::

    $ ln -s /mnt/flooding/Flooding/ror_keringen var


GISDATA
-------
Copy shape-files to ``BUILDOUT_DIR/var/gisdata`` from old-webserver.


EXCEL files
-----------
Copy excel-files to ``BUILDOUT_DIR/var/excel`` from old-webserver.


Setup mount to flod-share
-------------------------
Set ``cifspw``, mount in ``fstab``. Then create dir ``/mnt/flod-share``.

    $ sudo mkdir /mnt/flod-share
    $ sudo chown buildout:buildout /mnt/flod-share
    $ sudo mkdir -p /p-isilon-d1.external-nens.local/nens/flooding
    $ sudo chown buildout:buildout -R /p-isilon-d1.external-nens.local
    $ ln -s /mnt/flod-share flod-share
    $ ln -s /mnt/flod-share/pyramids pyramids
    $ ln -s /mnt/flod-share/ror_keringen ror_keringen
    $ ln -s /mnt/flod-share/exportruns/export_runs_csvs export_run_results

Symlink a buildout configuration
--------------------------------

Initially, there's no ``buildout.cfg``. You need to make that a symlink to the
correct configuration. On your development machine, that is
``development.cfg`` (and ``staging.cfg`` or ``production.cfg``, for instance
on the server)::

    $> ln -s development.cfg buildout.cfg


Raster Server
-------------

We also use an instance of the "raster-server" to serve WMS layers for
grid data. The grid data is stored as gislib "pyramids".

To use gislib and raster-server in Flooding, both need to be checked out
as development packages, using the "flooding-branch" branch.

Running Buildout, a configuration file for the raster-server is
created as etc/rasterserver.json. It says that the rasters are served
from BUILDOUT_DIR/var/pyramids. It is possible to symlink
/mnt/flooding/Flooding/pyramids to that directory, or to copy a few
rasters from the mounted share to that directory, or to change the
etc/rasterserver.json.in input file to use something file (in that
case, don't commit it).

The command to run the raster-server in development is, in the
buildout directory:


    $> RASTER_SERVER_SETTINGS=etc/rasterserver.json bin/runflask

The server will run at 0.0.0.0:5000 and visiting it should show a
working demo page where the available layers can be shown (although
there might be way too many for the page to render if you are using
the full Flooding share).

The URL used to find the WMS server is set in the Django settings as
RASTER_SERVER_URL. developmentsettings.py sets it to
'http://127.0.0.1:5000/wms' by default, change it to whatever you need
in localsettings.py if you are using virtual machines or similar.


Windows (task-server)
--------------------------------

* Check out the ``windows`` subdirectory, and customize it if needed.
* Check out the ``objectenbeheer/settings/windows.py`` module, and customize it if needed.

* Run ``build_windows.sh`` from Linux to wrap everything in a nice zip.

* In Windows, download Python 2.7.x from http://www.python.org/download/.
* In Windows, download Psycopg2 from http://www.stickpeople.com/projects/python/win-psycopg/.

* Extract the zip in the configured place, e.g. ``D:\Programs\flooding``.

* In Windows, configure your ``PYTHONPATH`` environment variable to point to the absolute path of the ``flooding\lib`` subdirectory.
  If you don't know how to do this, read https://kb.wisc.edu/cae/page.php?id=24500.

* To tune local settings like the database connection, create or edit ``objectenbeheer\lib\flooding\localsettings.py``.



Cleanup groupimport and importscenario
--------------------------------------
Run periodically ``cleanup_groupimport_dirs`` and ``cleanup_importscenario_dirs``
management command of ``flooding-lib`` package to remove wrong/unregistered 
import-files. The dirs locate in ``var/media/import/``


Symlinks on windows
-------------------------
To avoid the problem with symlink on windows put the file ``sitecustomize.py`` 
into site-packages directory of your Python installation. The file located in 
windows dir ``{buildout:directory}/windows``


Gebiedsdekkende kaarten
-----------------------
The app based on exporttool, used some parts of exporttool and located
in ``flooding_lib/tools/gdmapstool``

The app requires ``gdmapstool.change_gdmap`` permission.







CHANGES.txt
Changelog of flooding
===================================================


1.96 (unreleased)
-----------------

- Ping nens to 1.13

- Ping flooding_lib to 3.32, to 4.2, to 4.3, 4.5, 4.7, 4,8

- Import dir moved from var/media to /mnt/flod-share/import

- Excel dir moved from var/excel to /mnt/flod-share/excel

- Edit help file.

- Update raven

1.95.1 (2014-04-11)
-------------------

- Nothing changed yet.


1.95 (2014-04-11)
-----------------

- Add language settings.
- Remove 3di tasks from recipe of supervisor
- Remove references to raster-server, because it's a part of
  flooding-lib now.

1.94 (2014-02-10)
-----------------

- Add task 150, 180 to production-task-200.cfg

- Set paths to new share in settings.py.

1.93 (2014-01-29)
-----------------

- Put right branches of gislib and rasterserver into buildout config.


1.92 (2013-10-11)
-----------------

- Netcdf4 needs to be pinned at 1.0.4, incompatibility.


1.91 (2013-10-09)
-----------------

- Add settings for ror-wateren and ror-keringen
- Disable GZipMiddleware.
- Add path to mkdir recipe


1.90 (2013-09-09)
-----------------

- Updates:
    flooding-base to 1.40
      - Update smartclient to v9.0.
      - Add speed control button for animations
    flooding-lib to 2.92
      - Update templates to use smartclient v9.0.


1.89 (2013-08-14)
-----------------

- Updates
    flooding-base to 1.38
        Set projection to maplayer in config.js.


1.88 (2013-07-16)
-----------------

- Fix supervisor task call.


1.87 (2013-07-16)
-----------------

- Scedule tasks calculate_province_statistics_cronjob to run at 6:01 and
  create_province_excel_script at 6:30 daily.


1.86 (2013-07-15)
-----------------

- Unpinned flooding-lib. We got tired of keeping it up to date and
  updating these CHANGES. See the latest version of flooding-lib.


1.85 (2013-07-12)
-----------------

- Updates
    flooding-lib to 2.87
        - Cache the province a Scenario belongs to. Cache a province's
          scenario statistics.
	- Correctly create ROR approval Excel sheets for each province.
        - Don't generate them from the view if they already exist.
        - (In flooding) call the script each hour.
	- Fix error 'To many open files' by exporttool.

    flooding-base to 1.38
        - Fix OpenStreetMap layer.

1.84.2 (2013-07-10)
-------------------

- Nothing changed yet.


1.84.1 (2013-07-10)
-------------------

- Fix typo in supervisor command name


1.84 (2013-07-10)
-----------------

- Flooding-lib to 2.86 for:
    - Correctly create ROR approval Excel sheets for each province.
    - Don't generate them from the view if they already exist.
- Call the ROR Excel script every hour



1.83 (2013-07-08)
-----------------

- Flooding-lib to 2.85: added some caches to speed up the ROR
  dashboard page.  Also added a cronjob that runs every hour, using
  supervisor to make sure it doesn't start overlapping with itself.


1.82 (2013-07-03)
-----------------

- Updates
    flooding-lib to 2.84
        - Add field 'initialcrest' to ScenarioBreach models
          for import functionality.

1.81 (2013-07-02)
-----------------

- Updates
    flooding-lib to 2.83
        - Fix script calc.py for defining 'waterstandswerloop' points.

1.80 (2013-06-20)
-----------------

- Updates
    flooding-lib to 2.82
        - Set permission check on editing of export_run.


1.79 (2013-06-18)
-----------------

- Updates
    flooding-lib to 2.80, 2.81
        - Add functionality to manage initial_level for openbreach task
        - Override filse in model.zip for openbreach task
        - Check with pep8
        - Fix tests
    flooding-base to 1.37
        - Fix bug in NGeneral.js


1.78 (2013-06-12)
-----------------

Updates
    flooding-base to 1.36
        - Extended intervalFormatter to use an interval > 30 days.


1.77 (2013-05-21)
-----------------

- Extend request's timout to 600 sec. due error on 'flooding/shared/ror/'.


1.76 (2013-05-16)
-----------------

- Updates
    flooding-lib to 2.79
        - Close gdal dataset properly for exporttool.


1.75 (2013-05-16)
-----------------

- Updates
    flooding-lib to 2.78
        - Completed implementation of 'edit' functionality for export tool.


1.74 (2013-05-15)
-----------------

- Updates
    flooding-lib to 2.77
        - Revrite calculate_export_maps script to reduce memory usage.
        - Add functionality to edit exports (partly implemented).

1.73 (2013-05-07)
-----------------

- Updates
    flooding-lib to 2.74, 2.75
        - Remove creating .csv and .txt files on creating a new ror-export.
        - Use gtiff driver in place of mem to reduce memeory usage for
          ror-export calc


1.72 (2013-05-07)
-----------------

- New versions of example groupimport Excel sheets.


1.71 (2013-04-22)
-----------------

- Updates
    flooding-lib to 2.74, 2.75
        - Define transparency for all values that less
          then the lowest leftbound.
        - Import simplejson form django.utils


1.70 (2013-04-17)
-----------------

- Updates
    flooding-lib to 2.73:
        - Removed import unicode_literals from presentationlayer_generation.py
          due problem with gdal.


1.69 (2013-04-15)
-----------------

- Updates
    flooding-lib to 2.72:
        - Set max_uncompressed_zip_size to None for removing comment line
  	  from zips > 1GB.


1.68 (2013-04-09)
-----------------

- Updates
    flooding-lib to 2.71:
        - Fix bug in png_generation task by selecting colormapping.


1.67 (2013-04-09)
-----------------

- Updates
    flooding-lib to 2.70
        - Delete gdal's dataset to reduce memory usage
          for ror-export task.


1.66 (2013-04-09)
-----------------

- empty release.


1.65 (2013-04-08)
-----------------

- Repaire fabfile.cfg.


1.64 (2013-04-08)
-----------------

- Updates
    flooding-lib to 2.69:
        - More encoding fixes in exporttool.


1.63 (2013-04-08)
-----------------

- Updates
    flooding-lib to 2.68:

flooding/settings.py
# Base Django settings, suitable for production.  Imported (and partly
# overridden) by developmentsettings.py which also imports
# localsettings.py (which isn't stored in svn).
# Build19890150.1534fout takes care of using the correct one.  So:
# "DEBUG = TRUE" goes into developmentsettings.py and per-developer
# database ports go into localsettings.py.  May your hear turn purple
# if you ever put personal settings into this file or into
# developmentsettings.py!

import logging
import os
import tempfile
import pkg_resources
import matplotlib
matplotlib.use('Agg')

import sys
try:
    import PIL.Image
    sys.modules['Image'] = PIL.Image
except ImportError:
    import Image

from .settingshelper import setup_logging

from logging.handlers import RotatingFileHandler

# SETTINGS_DIR allows media paths and so to be relative to this settings file
# instead of hardcoded to c:\only\on\my\computer.
SETTINGS_DIR = os.path.dirname(os.path.realpath(__file__))

# BUILDOUT_DIR is for access to the "surrounding" buildout, for instance for
# BUILDOUT_DIR/var/media files to give django-staticfiles a proper place
# to place all collected static files.
BUILDOUT_DIR = os.path.abspath(os.path.join(SETTINGS_DIR, '..'))

# Downloadable Excel files
EXCEL_DIRECTORY = os.path.join(BUILDOUT_DIR, "var", "excel")

# Triple blast.  Needed to get matplotlib from barfing on the server: it needs
# to be able to write to some directory.
if 'MPLCONFIGDIR' not in os.environ:
    os.environ['MPLCONFIGDIR'] = tempfile.gettempdir()

# Production, so DEBUG is False. developmentsettings.py sets it to True.
DEBUG = False
# Show template debug information for faulty templates.  Only used when DEBUG
# is set to True.
TEMPLATE_DEBUG = True

# ADMINS get internal error mails, MANAGERS get 404 mails.
ADMINS = (
    # ('Your Name', 'your_email@domain.com'),
)
MANAGERS = ADMINS

DATABASES = {
    'default': {
        'ENGINE': 'django.contrib.gis.db.backends.postgis',
        'HOST': 'p-flod-db-00-d1.external-nens.local',
        'NAME': 'flooding',
        'USER': 'flooding',
        'PORT': 5432, # Required for mapnik
        'PASSWORD': 'XXXXX'  # See localproductionsettings.py
        }
    }


# Almost always set to 1.  Django allows multiple sites in one database.
SITE_ID = 1

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name although not all
# choices may be available on all operating systems.  If running in a Windows
# environment this must be set to the same as your system time zone.
TIME_ZONE = 'Europe/Amsterdam'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'nl'
# For at-runtime language switching.  Note: they're shown in reverse order in
# the interface!
ugettext = lambda s: s

LANGUAGES = (
    ('nl', ugettext('Nederlands')),
    ('en', ugettext('English')),
)

LOCALE_PATHS = (os.path.join(BUILDOUT_DIR, 'src', 'flooding-lib', 'flooding_lib', 'locale'),)
# If you set this to False, Django will make some optimizations so as not to
# load the internationalization machinery.
USE_I18N = True

# Absolute path to the directory that holds user-uploaded media.
MEDIA_ROOT = os.path.join(BUILDOUT_DIR, 'var', 'media')
# Absolute path to the directory where django-staticfiles'
# "bin/django build_static" places all collected static files from all
# applications' /media directory.
STATIC_ROOT = os.path.join(BUILDOUT_DIR, 'var', 'static')

STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
    # Enable support for django-compressor.
    'compressor.finders.CompressorFinder',
    )

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash if there is a path component (optional in other cases).
MEDIA_URL = '/media/'
# URL for the per-application /media static files collected by
# django-staticfiles.  Use it in templates like
# "{{ MEDIA_URL }}mypackage/my.css".
STATIC_URL = '/static_media/'
# URL prefix for admin media -- CSS, JavaScript and images. Make sure to use a
# trailing slash.  Uses STATIC_URL as django-staticfiles nicely collects
# admin's static media into STATIC_ROOT/admin.
ADMIN_MEDIA_PREFIX = STATIC_URL + 'admin/'

# Storage engine to be used during compression
###COMPRESS_STORAGE = "staticfiles.storage.StaticFileStorage"

# The URL that linked media will be read from and compressed media will be
# written to.
###COMPRESS_URL = STATIC_URL
# The absolute file path that linked media will be read from and compressed
# media will be written to.
###COMPRESS_ROOT = STATIC_ROOT

# django-staticfiles needs an extra context processor to allow you to use {{
# STATIC_URL }}myapp/my.css in your templates.
TEMPLATE_CONTEXT_PROCESSORS = (
    # Default items.
    'django.contrib.auth.context_processors.auth',
    'django.core.context_processors.debug',
    'django.core.context_processors.i18n',
    'django.core.context_processors.media',
    'django.core.context_processors.static')

MIDDLEWARE_CLASSES = (
    # Gzip needs to be at the top.
    #'django.middleware.gzip.GZipMiddleware',
    # Below is the default list, don't modify it.
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    #'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    )

# Make this unique, and don't share it with anybody.
SECRET_KEY = 'b8cdceb3-4879-4dcc-96ef-5030c2493fa0'

ROOT_URLCONF = 'flooding.urls'

CACHE_BACKEND = 'file://%s' % os.path.join(BUILDOUT_DIR, 'var', 'cache')
# Note: for development only, check django website for caching solutions for
# production environments
# ^^^ TODO

INSTALLED_APPS = (
    'flooding',
    'flooding_lib.tools.pyramids',
    'flooding_lib.sharedproject',
    'flooding_base',
    'flooding_presentation',
    'flooding_visualization',  # Must be below flooding_presentation
    'flooding_lib.tools.approvaltool',
    'lizard_worker',
    'lizard_raster',
    'flooding_lib',  # Must be below flooding_visualization,
                     # flooding_presentation, lizard_worker and
                     # approvaltool
    'flooding_lib.tools.importtool',
    'flooding_lib.tools.exporttool',
    'django.contrib.staticfiles',
    'raven.contrib.django.raven_compat',
    'south',
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.gis',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django_extensions',
    'django_nose',  # Must be below south
    'supervisor',
    'gunicorn',
    'markdown_deux',  # For markdown template filter in flooding-base
)

TEST_RUNNER = 'django_nose.NoseTestSuiteRunner'

LOGGING = setup_logging(
    BUILDOUT_DIR, console_level=None, file_level='WARN')

# We create a handler to be able to show the tail of the Django log to the
# user. The handler implements the tail through up to two log files that are
# each up to 4 KB large. The log files are called
#
#    - /var/log/django_tail.log and
#    - /var/log/django_tail.log.1
#
# The latter log file is only created when the former has reached its maximum
# size.
TAIL_LOG = os.path.join(BUILDOUT_DIR, 'var', 'log', 'django_tail.log')

handler = RotatingFileHandler(TAIL_LOG, maxBytes=4096, backupCount=1)
logging.getLogger().addHandler(handler)

SYMBOLS_DIR = pkg_resources.resource_filename(
    'flooding_visualization', 'media/flooding_visualization/symbols')
FLOODING_SHARE = '//p-isilon-d1.external-nens.local/nens/flooding/flod-share/'
EXTERNAL_PRESENTATION_MOUNTED_DIR = os.path.join(
    FLOODING_SHARE, 'presentationdatabase_totaal')
EXTERNAL_RESULT_MOUNTED_DIR = os.path.join(
    FLOODING_SHARE, 'resultaten')
# Set /tmp on separate divice and check it periodically
TMP_DIR = '/tmp'

GIS_DIR = os.path.join(BUILDOUT_DIR, 'var', 'gisdata')

RASTER_SERVER_URL = "http://flooding.lizard.net/wms"

#location of directories for task execution. Pelase configure to local
#installation root of HIS schade en slachtoffers module
HISSSM_ROOT = ''
#root of sobek program installation
SOBEK_PROGRAM_ROOT = ''
#root of sobek projects
SOBEK_PROJECT_ROOT = ''
#root of temporary directory for flooding tasks on windows
TMP_ROOT = 'c:\\temp'

PERFORM_TASK_MODULE = "flooding_lib.tasks.perform_task"
PERFORM_TASK_FUNCTION = "perform_task"

#queue's setting for flooding-worke
QUEUES = {
    "default": {
        "exchange": "",
        "binding_key": "default"},
    "logging": {
        "exchange": "router",
        "binding_key": "logging"},
    "failed": {
        "exchange": "router",
        "binding_key": "failed"},
    "sort": {
        "exchange": "router",
        "binding_key": "sort"},
    "120": {
        "exchange": "router",
        "binding_key": "120"},
    "130": {
        "exchange": "router",
        "binding_key": "130"},
    "132": {
        "exchange": "router",
        "binding_key": "132"},
    "134": {
        "exchange": "router",
        "binding_key": "134"},
    "150": {
        "exchange": "router",
        "binding_key": "150"},
    "155": {
        "exchange": "router",
        "binding_key": "155"},
    "160": {
        "exchange": "router",
        "binding_key": "160"},
    "162": {
        "exchange": "router",
        "binding_key": "162"},
    "180": {
        "exchange": "router",
        "binding_key": "180"},
    "185": {
        "exchange": "router",
        "binding_key": "185"},
    "190": {
        "exchange": "router",
        "binding_key": "190"},
    "200": {
        "exchange": "router",
        "binding_key": "200"},
    "210": {
        "exchange": "router",
        "binding_key": "210"},
    "220": {
        "exchange": "router",
        "binding_key": "220"},
    "900": {
        "exchange": "router",
        "binding_key": "900"},
}

HEARTBEAT_QUEUES = ["120", "130", "132", "134", "150", "155", "160", "162", "180", "185", "190", "200", "210", "220"]

# TODO: configure your broker settings
BROKER_SETTINGS = {
    "BROKER_HOST": "localhost",
    "BROKER_PORT": 5672,
    "BROKER_USER": "",
    "BROKER_PASSWORD": "",
    "BROKER_VHOST": "TODO",
    "HEARTBEAT": False
}

# import ror-keringen
ROR_KERINGEN_PATH = os.path.join(BUILDOUT_DIR, 'var', 'ror_keringen')
ROR_KERINGEN_APPLIED_PATH = os.path.join(ROR_KERINGEN_PATH, 'applied')
ROR_KERINGEN_NOTAPPLIED_PATH = os.path.join(ROR_KERINGEN_PATH, 'not_applied')

RAVEN_CONFIG = {}

SESSION_SERIALIZER = 'django.contrib.sessions.serializers.PickleSerializer'

try:
    from flooding.localproductionsettings import *
    # For local production overrides (DB passwords, for instance)
except ImportError:
    pass
